### PR TITLE
[WIP] Issue 292

### PR DESCRIPTION
**Pull request by antoyo**
Hi.
I started to work on issue #292 and I was able to make more hints working, but I think my solution is too complicated. There must be a simpler way.
My current solution is to be tested, but it seems most of the links now works, except perhaps in `iframe`.

As an alternative, I thought of using the top right corner of the `QRect` returned by `geometry()` but it only fixes some hints, not all.

Do you have any idea?

### CODE REVIEW

**Comment by The-Compiler** [comment]
Phew! :smile:

It's certainly possible a solution which works is a bit complicated - it also took me a while until the current solution was working properly :smile: - this is also why I'd like to write some tests for corner-cases before changing anything in this area.

One thing I don't really get is why you're doing this selection hack - what's the point of selecting all children and then getting the first element?

It also seems `_get_element_absolute_geometry` basically does what `rect_on_view` does, just not recursively (and not considering the scroll position) - did you have trouble with `rect_on_view`?

**Comment by antoyo** [comment]
I was doing the selection hack to get the position of the first character instead of the position of the whole rect to get the hint at the right position when the link is text wrapped on multiple lines. It also allowed me to ignore the `border-radius`.
I found an alternative, so I removed this hack.
I cannot use `rect_on_view` because it relies on `geometry()` which gives me the rectangle around the whole text of a link, which is the wrong rect when the text is wrapped on multiple lines.
In my latest commits, I fixed the issue with `iframes`.

I agree that we need some tests for the corner cases, but I believe we'll need to find more corner cases.

**Comment by The-Compiler** [comment]
I was about to suggest `getClientRects` :smile:

You can pass in an `elem_geometry` to `rect_on_view`, then it will use that instead of `elem.geometry()`.

File: qutebrowser/browser/hints.py [hunk 1]
```diff
@@ -355,6 +356,90 @@ def _set_style_properties(self, elem, label):
...
+            if show:
+                body = elem.document().findFirst('body')
+                if(body.styleProperty('position', QWebElement.ComputedStyle) ==
+                        'relative'):
+                    body_rect = elem.evaluateJavaScript(
+                        'document.body.getBoundingClientRect()'
+                    )
+                    rect.translate(-body_rect['left'], -body_rect['top'])
+                    frame = elem.webFrame()
+                    rect.translate(-frame.scrollPosition())
```
Comment (The-Compiler) [inline qutebrowser/browser/hints.py]: It might make sense to play with other [position: values](http://www.w3schools.com/cssref/pr_class_position.asp) for the hint CSS - maybe using `fixed` instead of `absolute` would help?

Comment (antoyo) [inline qutebrowser/browser/hints.py]: Vimperator creates an element (with `fixed` position) containing all the hints.
The issue with this is that the elements stay `fixed` if we scroll, but maybe it is not a real problem.
Do you see any issue with my solution?

Comment (The-Compiler) [inline qutebrowser/browser/hints.py]: I don't see an issue - I just want to avoid more complexity if it can be avoided :wink:

I hope I'll get to writing some tests for hint positioning soon. I definitely have some corner cases in mind.

**Comment by The-Compiler** [comment]
I merged #1311 by @lahwaacz now, which I think obsoletes this PR and is a bit simpler - thanks either way!

@antoyo If you see something which is handled by your PR but not by #1311, please let me know!

**Comment by antoyo** [comment]
@The-Compiler As far as I can see, only the buttons on https://duckduckgo.com/settings are not clickable with the other pull request while they are with mine. But I think I had to add a special case for this page as the input checkboxes are 1px by 1px.
